crates/bevy_platform_support/src/staging.rs

```diff
@@ -0,0 +1,460 @@
+//! Provides an abstracted system for staging modifications to data structures that rarely change. See [`StageOnWrite`] as a starting point.
+
+use core::ops::Deref;
+
+use crate::alloc::sync::Arc;
+use crate::sync::{RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+/// Signifies that this type represents staged changes to [`Cold`](Self::Cold).
+pub trait StagedChanges: Default {
+    /// The more compact data structure that these changes compact into.
+    type Cold;
+
+    /// This applies these changes to the passed [`Cold`](Self::Cold). When this is done, there should be no more staged changes, and [`any_staged`](Self::any_staged) should be false.
+    fn apply_staged(&mut self, storage: &mut Self::Cold);
+
+    /// Returns true if and only if there are staged changes that could be applied.
+    fn any_staged(&self) -> bool;
+}
+
+/// A trait that signifies that it holds an immutable reference to a cold type (ie. [`StagedChanges::Cold`]).
+pub trait ColdStorage<T: StagedChanges>: Deref<Target = T::Cold> {}
+
+/// A struct that allows staging changes while reading from cold storage. Generally, staging changes should be implemented on this type.
+pub struct Stager<'a, T: StagedChanges> {
+    /// The storage that is read optimized.
+    pub cold: &'a T::Cold,
+    /// The staged changes.
+    pub staged: &'a mut T,
+}
+
+/// A struct that allows accessing changes while reading from cold storage. Generally, reading data should be implemented on this type.
+#[derive(Copy)]
+pub struct StagedRef<'a, T: StagedChanges> {
+    /// The storage that is read optimized.
+    pub cold: &'a T::Cold,
+    /// The staged changes.
+    pub staged: &'a T,
+}
+
+/// A locked version of [`Stager`]. Use this to hold a lock guard while using [`StagerLocked::as_stager`] or similar.
+pub struct StagerLocked<'a, T: StagedChanges, C: ColdStorage<T>> {
+    cold: C,
+    staged: RwLockWriteGuard<'a, T>,
+}
+
+/// A locked version of [`StagedRef`] Use this to hold a lock guard while using [`StagerLocked::as_staged_ref`].
+pub struct StagedRefLocked<'a, T: StagedChanges, C: ColdStorage<T>> {
+    cold: C,
+    staged: RwLockReadGuard<'a, T>,
+}
+
+/// A general purpose enum for representing data that may or may not need to be staged.
+pub enum MaybeStaged<C, S> {
+    /// There is no staging necessary.
+    Cold(C),
+    /// There is staging necessary.
+    Staged(S),
+}
+
+/// A struct that allows read-optimized operations while still allowing mutation. When mutattions are made,
+/// they are staged. Then, at user-defined times, they are applied to the read-optimized storage. This allows muttations
+/// through [`RwLock`]s without needing to constantly lock old or cold data.
```

Comment:
```suggestion
/// A struct that allows read-optimized operations while still allowing mutation.
/// When mutations are made, they are staged.
/// Then, at user-defined times, they are applied to the read-optimized storage.
/// This allows mutations through [`RwLock`]s without needing to constantly lock old or cold data.
```
Simple typo and reformatting for new lines. In Rust documentation, you can put new sentences in the same paragraph on the next documentation line. I personally find that easier to read, and it's usually more friendly for Git diffs.